File: webserver/src/main/java/com/llh/webserver/service/impl/AuthServiceImpl.java
package com.llh.webserver.service.impl;

import cn.hutool.core.bean.BeanUtil;
import com.llh.webserver.common.constant.JWTConstant;
import com.llh.webserver.common.exception.BusinessException;
import com.llh.webserver.common.exception.msg.BasicExpEnum;
import com.llh.webserver.model.SysUser;
import com.llh.webserver.pojo.auth.AuthTokenVO;
import com.llh.webserver.pojo.auth.LoginVO;
import com.llh.webserver.pojo.auth.RegisterOrUpdateVO;
import com.llh.webserver.service.AuthService;
import com.llh.webserver.util.jwt.JwtTokenUtil;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import javax.transaction.Transactional;
import java.util.HashMap;
import java.util.Map;

/**
 * AuthServiceImpl
 * <p>
 * CreatedAt: 2020-04-18 18:09
 *
 * @author llh
 */
@Service("authService")
@Slf4j
public class AuthServiceImpl extends SysUserServiceImpl implements AuthService {
    @Autowired
    private PasswordEncoder passwordEncoder;

    @Override
    public AuthTokenVO login(LoginVO loginVO) {
        SysUser details = (SysUser) loadUserByUsername(loginVO.getUsername());
        if (!passwordEncoder.matches(loginVO.getPassword(), details.getPassword())) {
            log.info("密码不正确 : {} ", loginVO);
            throw new BusinessException(BasicExpEnum.Auth_ERROR);
        }
        return createAccountTokenVO(details);
    }

    @Override
    @Transactional
    public RegisterOrUpdateVO register(RegisterOrUpdateVO registerVO) {
        SysUser user = new SysUser();
        BeanUtil.copyProperties(registerVO, user, getCopyOptions());
        user.setPassword(registerVO.getPassword());
        save(user);
        return registerVO;
    }

    @Override
    public boolean logout(AuthTokenVO tokenVO) {
        return false;
    }


    /**
     * 生成token信息类
     *
     * @param userDetails 用户信息
     * @return 帐户token信息类
     */
    private AuthTokenVO createAccountTokenVO(SysUser userDetails) {
        AuthTokenVO tokenVO = new AuthTokenVO();
        Map<String, Object> claims = genClaims(userDetails);
        tokenVO.setAccessToken(genAccessToken(userDetails.getId(), claims))
            .setRefreshToken(genRefreshToken(userDetails.getId(), claims))
            .setId(userDetails.getId())
            .setUsername(userDetails.getUsername());
        return tokenVO;
    }

    /**
     * 设置基本信息
     *
     * @param user 用户信息
     * @return 用于填充JWT中claims部分信息的Map
     */
    private Map<String, Object> genClaims(SysUser user) {
        Map<String, Object> claims = new HashMap<>();
        claims.put(JWTConstant.JWT_USERNAME, user.getUsername());
        claims.put(JWTConstant.JWT_ROLES_KEY, user.getAuthorities());
        claims.put(JWTConstant.JWT_USER_ID, user.getId());
        return claims;
    }

    /**
     * 生成访问token
     *
     * @param userId 用户id
     * @param claims JWT中claims部分信息
     * @return 访问token/接入令牌
     */
    private String genAccessToken(String userId, Map<String, Object> claims) {
        return JwtTokenUtil.generateAccessToken(userId, claims);
    }


    /**
     * 生成刷新token
     *
     * @param userId 用户id
     * @param claims JWT中claims部分信息
     * @return 刷新token
     */
    private String genRefreshToken(String userId, Map<String, Object> claims) {
        return JwtTokenUtil.generateRefreshToken(userId, claims);
    }
}

File: server4kt/src/main/kotlin/com/llh/server/service/sys/impl/AuthServiceImpl.kt
package com.llh.server.service.sys.impl

import com.llh.server.common.constant.AuthExpInfo
import com.llh.server.common.exception.BasicException
import com.llh.server.common.util.JWT_ROLES_KEY
import com.llh.server.common.util.JWT_USERNAME
import com.llh.server.common.util.JwtTokenUtil
import com.llh.server.pojo.AccountVO
import com.llh.server.pojo.vo.AuthTokenVO
import com.llh.server.pojo.vo.LoginVO
import com.llh.server.service.sys.AuthService
import com.llh.server.service.sys.SysUserService
import org.apache.logging.log4j.kotlin.Logging
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service

/**
 * AuthServiceImpl
 *
 * CreatedAt: 2020-06-06 15:50
 *
 * @author llh
 */
@Service("authService")
class AuthServiceImpl : AuthService, Logging {
    @Autowired
    private lateinit var passwordEncoder: PasswordEncoder

    @Autowired
    private lateinit var jwtTokenUtil: JwtTokenUtil

    @Autowired
    @Qualifier("sysUserService")
    private lateinit var sysUserService: SysUserService

    override fun login(loginVO: LoginVO): AuthTokenVO {
        val details = sysUserService
            .loadUserByUsername(loginVO.username) as AccountVO
        if (!passwordEncoder.matches(loginVO.password, details.password)) {
            logger.info("$loginVO 密码不正确.")
            throw BasicException(AuthExpInfo.LOGIN_ERROR)
        }
        return createAccountTokenVO(details)
    }

    /**
     * 生成token信息类
     */
    private fun createAccountTokenVO(details: AccountVO): AuthTokenVO {
        val claims = genClaims(details)
        return AuthTokenVO(
            accessToken = genAccessToken(details.id, claims),
            refreshToken = genRefreshToken(details.id, claims),
            username = details.username
        )
    }

    /**
     * 设置基本信息
     */
    private fun genClaims(user: UserDetails): MutableMap<String, Any> {
        val claims = mutableMapOf<String, Any>()
        claims[JWT_USERNAME] = user.username
        claims[JWT_ROLES_KEY] = user.authorities
        return claims
    }

    /**
     * 生成访问token
     */
    private fun genAccessToken(userId: String, claims: Map<String, Any>): String {
        return jwtTokenUtil.generateAccessToken(userId, claims)
    }

    /**
     * 生成刷新token
     */
    private fun genRefreshToken(userId: String, claims: Map<String, Any>): String {
        return jwtTokenUtil.generateRefreshToken(userId, claims)
    }
}
